### PR TITLE
feat: Functions sdk update

### PR DESCRIPTION
**Pull request by sfc-gh-asawicki**
Update and test functions SDK as part of ongoing functions&procedures rework:
- Regenerate SDK after adjusting defs
- Wrap function definitions in `$$`
- Wrap arguments in double quotes
- Extract FunctionDetails out of DESCRIBE output rows
- Add MetricLevel to the SDK
- Adjust alter's structure (add set/unset)
- Add missing fields to SHOW output
- Use extended in for SHOW
- Add/Generate assertions for function, function details, and function parameters
- Add missing unit tests and adjust existing ones
- Add missing integration tests and remove most of the existing ones or adjust them
- Adjust existing resource

Additionally:
- Allow simplified setup for local testing
- Add helpers to interact with user's stage
- Improve SDK generation (remove part of the unnecessary newlines - some are still left)

### CODE REVIEW

File: pkg/acceptance/testdatatypes/testdatatypes.go [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package testdatatypes
+
+import "github.com/Snowflake-Labs/terraform-provider-snowflake/pkg/sdk/datatypes"
+
+var (
+	DataTypeNumber_36_2, _ = datatypes.ParseDataType("NUMBER(36, 2)")
```
Comment (sfc-gh-jmichalak) [inline pkg/acceptance/testdatatypes/testdatatypes.go:6]: Nit: we can introduce something like `MustParseDataType`, similar to `regexp.MustCompile`

File: pkg/sdk/testint/functions_integration_test.go [hunk 2]
```diff
@@ -281,89 +1381,175 @@ func TestInt_OtherFunctions(t *testing.T) {
...
-		f := createFunctionForSQLHandle(t, true, true)
+		assertions.AssertThatObject(t, objectassert.Function(t, id).
+			HasName(id.Name()).
+			HasDescription("new comment"),
+		)
 
-		id := f.ID()
-		err := client.Functions.Alter(ctx, sdk.NewAlterFunctionRequest(id).WithUnsetTraceLevel(true))
+		assertions.AssertThatObject(t, objectassert.FunctionDetails(t, id).
+			HasExternalAccessIntegrations(fmt.Sprintf(`[%s]`, externalAccessIntegration.FullyQualifiedName())).
```
Comment (sfc-gh-jmichalak) [inline pkg/sdk/testint/functions_integration_test.go:1423]: Can we have an extension function for these types and pass slices/maps there?

**Review by sfc-gh-jmichalak — CHANGES_REQUESTED** [review]

File: pkg/acceptance/bettertestspoc/assert/objectparametersassert/gen/object_parameters_def.go [hunk 3]
```diff
@@ -205,4 +205,15 @@ var allObjectsParameters = []SnowflakeObjectParameters{
 			// TODO(SNOW-1348092 - next prs): Add parameters
 		},
 	},
+	{
+		Name:   "Function",
+		IdType: "sdk.SchemaObjectIdentifierWithArguments",
+		Level:  sdk.ParameterTypeFunction,
+		Parameters: []SnowflakeParameter{
+			{ParameterName: string(sdk.FunctionParameterEnableConsoleOutput), ParameterType: "bool", DefaultValue: "false", DefaultLevel: "sdk.ParameterTypeSnowflakeDefault"},
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/acceptance/bettertestspoc/assert/objectparametersassert/gen/object_parameters_def.go:213]: Yes, exactly

File: pkg/acceptance/helpers/function_client.go [hunk 4]
```diff
@@ -33,27 +36,108 @@ func (c *FunctionClient) CreateWithIdentifier(t *testing.T, id sdk.SchemaObjectI
...
+	id := c.ids.RandomSchemaObjectIdentifierWithArguments()
+
+	definition := c.SampleSqlDefinition(t)
+	dt := sdk.NewFunctionReturnsResultDataTypeRequest(dataType)
+	returns := sdk.NewFunctionReturnsRequest().WithResultDataType(*dt)
+	request := sdk.NewCreateForSQLFunctionRequestDefinitionWrapped(id.SchemaObjectId(), *returns, definition)
+
+	err := c.client().CreateForSQL(ctx, request)
+	require.NoError(t, err)
+	t.Cleanup(c.DropFunctionFunc(t, id))
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/acceptance/helpers/function_client.go:104]: It's either added already or the function returns the cleanup itself (we have issue to clean this up in all the helpers: SNOW-1763610)

File: pkg/sdk/functions_gen_test.go [hunk 5]
```diff
@@ -804,16 +859,16 @@ func TestFunctions_CreateForScala(t *testing.T) {
...
 		opts.Comment = String("comment")
 		opts.Imports = []FunctionImport{
 			{
 				Import: "@udf_libs/echohandler.jar",
 			},
 		}
 		opts.Handler = "Echo.echoVarchar"
-		opts.FunctionDefinition = String("return x")
-		assertOptsValidAndSQLEquals(t, opts, `CREATE OR REPLACE TEMPORARY SECURE FUNCTION %s (x VARCHAR(100) DEFAULT 'test') COPY GRANTS RETURNS VARCHAR(100) NOT NULL LANGUAGE SCALA CALLED ON NULL INPUT IMMUTABLE RUNTIME_VERSION = '2.0' COMMENT = 'comment' IMPORTS = ('@udf_libs/echohandler.jar') HANDLER = 'Echo.echoVarchar' AS 'return x'`, id.FullyQualifiedName())
+		opts.FunctionDefinition = String(wrapFunctionDefinition("return x"))
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/functions_gen_test.go:870]: I will add it

File: pkg/sdk/common_types.go [hunk 6]
```diff
@@ -233,7 +233,7 @@ func NullInputBehaviorPointer(v NullInputBehavior) *NullInputBehavior {
 
 const (
 	NullInputBehaviorCalledOnNullInput NullInputBehavior = "CALLED ON NULL INPUT"
-	NullInputBehaviorReturnNullInput   NullInputBehavior = "RETURN NULL ON NULL INPUT"
+	NullInputBehaviorReturnNullInput   NullInputBehavior = "RETURNS NULL ON NULL INPUT"
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/common_types.go:236]: Will do

File: pkg/sdk/functions_def.go [hunk 7]
```diff
@@ -187,7 +202,7 @@ var FunctionsDef = g.NewInterface(
 		SQL("LANGUAGE SCALA").
 		PredefinedQueryStructField("NullInputBehavior", "*NullInputBehavior", g.KeywordOptions()).
 		PredefinedQueryStructField("ReturnResultsBehavior", "*ReturnResultsBehavior", g.KeywordOptions()).
-		OptionalTextAssignment("RUNTIME_VERSION", g.ParameterOptions().SingleQuotes()).
+		TextAssignment("RUNTIME_VERSION", g.ParameterOptions().SingleQuotes().Required()).
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/functions_def.go:205]: Yes, there is a TODO left in functions_integration_test.go: `// TODO [SNOW-1850370]: test confirming that runtime version is required for Scala function`

File: pkg/resources/function_acceptance_test.go [hunk 8]
```diff
@@ -142,7 +142,7 @@ func TestAcc_Function_complex(t *testing.T) {
 					resource.TestCheckResourceAttr(resourceName, "comment", "Terraform acceptance test"),
 					resource.TestCheckResourceAttr(resourceName, "statement", statement),
 					resource.TestCheckResourceAttr(resourceName, "arguments.#", "1"),
-					resource.TestCheckResourceAttr(resourceName, "arguments.0.name", "D"),
+					resource.TestCheckResourceAttr(resourceName, "arguments.0.name", "d"),
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/resources/function_acceptance_test.go:145]: Will do in the PR with resource rework.

File: pkg/resources/function.go [hunk 9]
```diff
@@ -571,7 +573,7 @@ func ReadContextFunction(ctx context.Context, d *schema.ResourceData, meta inter
 		switch desc.Property {
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/resources/function.go:573]: Will do

File: pkg/sdk/functions_gen_test.go [hunk 10]
```diff
@@ -253,16 +266,16 @@ func TestFunctions_CreateForJava(t *testing.T) {
...
+		opts.FunctionDefinition = String(wrapFunctionDefinition("return id + name;"))
+		assertOptsValidAndSQLEquals(t, opts, `CREATE OR REPLACE TEMPORARY SECURE FUNCTION %s ("id" NUMBER(36, 2), "name" VARCHAR(100) DEFAULT 'test') COPY GRANTS RETURNS TABLE ("country_code" VARCHAR(100), "country_name" VARCHAR(100)) NOT NULL LANGUAGE JAVA CALLED ON NULL INPUT IMMUTABLE RUNTIME_VERSION = '2.0' COMMENT = 'comment' IMPORTS = ('@~/my_decrement_udf_package_dir/my_decrement_udf_jar.jar') PACKAGES = ('com.snowflake:snowpark:1.2.0') HANDLER = 'TestFunc.echoVarchar' EXTERNAL_ACCESS_INTEGRATIONS = ("ext_integration") SECRETS = ('variable1' = %s, 'variable2' = %s) TARGET_PATH = '@~/testfunc.jar' AS $$return id + name;$$`, id.FullyQualifiedName(), secretId.FullyQualifiedName(), secretId2.FullyQualifiedName())
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/functions_gen_test.go:278]: Also, we will later add option to the SQL builder (fourth "quoting" option - double dollars; so this logic will be cleaned)

File: pkg/sdk/functions_ext.go [hunk 11]
```diff
@@ -1,5 +1,149 @@
...
+// FunctionDetails contains aggregated describe results for the given function.
+type FunctionDetails struct {
+	Signature                  string  // present for all function types
+	Returns                    string  // present for all function types
+	Language                   string  // present for all function types
+	Body                       *string // present for all function types (hidden when SECURE)
+	NullHandling               *string // present for all function types but SQL
+	Volatility                 *string // present for all function types but SQL
+	ExternalAccessIntegrations *string // list present for python, java, and scala
+	Secrets                    *string // map present for python, java, and scala
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/sdk/functions_ext.go:25]: yes, that's the plan, most probably when adding the resources (this change is already inflated :( ) - there is a TODO left in functions_integration_test

File: pkg/acceptance/helpers/function_client.go [hunk 12]
```diff
@@ -81,3 +165,81 @@ func (c *FunctionClient) DropFunctionFunc(t *testing.T, id sdk.SchemaObjectIdent
...
+			return %[3]s
+		}
+	}
+`, className, funcName, argName)
+}
+
+func (c *FunctionClient) SampleSqlDefinition(t *testing.T) string {
+	t.Helper()
+
+	return "3.141592654::FLOAT"
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/acceptance/helpers/function_client.go:238]: I will add a TODO for later to do this.

File: pkg/acceptance/helpers/function_setup_helpers.go [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+	t.Cleanup(c.Function.DropFunctionFunc(t, id))
+
+	// using os.CreateTemp underneath - last * in pattern is replaced with random string
+	modulePattern := fmt.Sprintf("example*%s.py", random.AlphaLowerN(3))
+	modulePath := c.Stage.PutOnUserStageWithContent(t, modulePattern, definition)
+	moduleFileName := filepath.Base(modulePath)
+
+	return &TmpFunction{
+		FunctionId: id,
+		ModuleName: strings.ReplaceAll(moduleFileName, ".py", ""),
```
[the remark this answers is not in the document]
Comment (sfc-gh-asawicki) [reply pkg/acceptance/helpers/function_setup_helpers.go:86]: Will do

File: pkg/acceptance/testdatatypes/testdatatypes.go (hunk 1 above)
Comment (sfc-gh-asawicki) [reply pkg/acceptance/testdatatypes/testdatatypes.go:6]: We have 	SNOW-1843440 for it. I will add here the same TODO from other part of our codebase: `	// TODO [SNOW-1843440]: create using constructors (when we add them)?`

File: pkg/sdk/testint/functions_integration_test.go (hunk 2 above)
Comment (sfc-gh-asawicki) [reply pkg/sdk/testint/functions_integration_test.go:1423]: Yes, I will add them

**Review by sfc-gh-jmichalak — APPROVED** [review]

**Pull request merged**